File: 543-DiameterofBinaryTree/543-DiameterofBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
   let diameter = 0
   dfs = (node) => {
    if(node == null) return 0;
    
    let left = dfs(node.left)
    let right = dfs(node.right)

    diameter = Math.max(diameter, left + right)

    return 1 + Math.max(left, right)
   }
   dfs(root)
   return diameter;
}

File: 543-DiameterofBinaryTree/543-DiameterofBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        self.diameter = 0
        def dfs(node):
            #base case
            if node == None: return 0

            left = dfs(node.left)
            right = dfs(node.right)

            #updating the diameter
            self.diameter = max(self.diameter, left + right)

            #returning the diameter
            return 1 + max( left, right)
        
        dfs(root)
        return self.diameter